lnwallet/channel.go

```diff
@@ -3740,34 +3901,139 @@ func (lc *LightningChannel) ReceiveRevocation(revMsg *lnwire.RevokeAndAck) ([]*P
 		// locked-in at this new state. By doing this we ensure that we
 		// don't re-forward any already processed HTLC's after a
 		// restart.
-		if htlc.EntryType == Add &&
-			remoteChainTail == htlc.addCommitHeightRemote &&
-			localChainTail >= htlc.addCommitHeightLocal {
+		if pd.EntryType == Add &&
+			remoteChainTail == pd.addCommitHeightRemote &&
+			localChainTail >= pd.addCommitHeightLocal {
 
-			htlc.isForwarded = true
-			htlcsToForward = append(htlcsToForward, htlc)
+			pd.SourceRef = &channeldb.AddRef{
+				Height: remoteChainTail,
+				Index:  addIndex,
+			}
+			addIndex++
+
+			pd.isForwarded = true
+			addsToForward = append(addsToForward, pd)
+		} else if pd.EntryType != Add &&
+			remoteChainTail >= pd.removeCommitHeightRemote &&
+			localChainTail >= pd.removeCommitHeightLocal {
+
+			pd.DestRef = &channeldb.SettleFailRef{
+				Source: lc.ShortChanID(),
+				Height: remoteChainTail,
+				Index:  settleFailIndex,
+			}
+			settleFailIndex++
+
+			pd.isForwarded = true
+			settleFailsToForward = append(settleFailsToForward, pd)
+		} else {
 			continue
 		}
 
-		if htlc.EntryType != Add &&
-			remoteChainTail >= htlc.removeCommitHeightRemote &&
-			localChainTail >= htlc.removeCommitHeightLocal {
+		// Knowing that this update is a part of this new commitment,
+		// we'll create a log update and note it's index in the log so
+		// we can later restore it properly if a restart occurs.
+		logUpdate := channeldb.LogUpdate{
+			LogIndex: pd.LogIndex,
+		}
 
-			htlc.isForwarded = true
-			htlcsToForward = append(htlcsToForward, htlc)
-			continue
+		// We'll map the type of the PaymentDescriptor to one of the
+		// four messages that it corresponds to and separate the updates
+		// into Adds and Settle/Fail/MalformedFail such that they can be
+		// written in the forwarding package.
+		switch pd.EntryType {
+		case Add:
+			htlc := &lnwire.UpdateAddHTLC{
+				ChanID:      chanID,
+				ID:          pd.HtlcIndex,
+				Amount:      pd.Amount,
+				Expiry:      pd.Timeout,
+				PaymentHash: pd.RHash,
+			}
+			copy(htlc.OnionBlob[:], pd.OnionBlob)
+			logUpdate.UpdateMsg = htlc
+			addUpdates = append(addUpdates, logUpdate)
```

Comment:
If the entry does not satisfy the conditions above (e.g. `pd.EntryType == Add && remoteChainTail == pd.addCommitHeightRemote && localChainTail >= pd.addCommitHeightLocal`) it will not be set `isForwarded=true`, and end up being added to the `fwdPkg` more than once. Is this intentional?

Reply:
I thinkk if that check fails (and the one after), we will just `continue` and wait until the condition is met? I may be misinterpreting this though

Reply:
yup reading it again, it looks like you are right. I missed the `continue`. 

This block of code starts to become a little bit confusing though. Is it possible to combine the above `if` statements with the `switch` somehow, to reduce the number of paths the execution can take?

Reply:
Alright, i've taken a stab at rewriting this section, you can let me know what you think after I rebase and push. Before I was trying to leave as much of the constraint validation intact as I a could, though I think this new version is a good compromise :)

Reply:
I'm good with this compromise 👍

Reply:
Woohoo! 🎉